libavcodec/qsv.c

```diff
@@ -103,17 +106,33 @@ static const struct {
 #undef MAP
 };
 
-int ff_qsv_profile_to_mfx(enum AVCodecID codec_id, int profile)
+int ff_qsv_profile_to_mfx(AVCodecContext *avctx)
 {
+    const AVCodecDescriptor *codec_desc;
+    int codec_id = avctx->codec_id;
+    int profile = avctx->profile;
     int i;
-    if (profile == FF_PROFILE_UNKNOWN)
+
+    codec_desc = avcodec_descriptor_get(avctx->codec_id);
+    if (!codec_desc) {
         return MFX_PROFILE_UNKNOWN;
+    }
 
     for (i = 0; i < FF_ARRAY_ELEMS(qsv_profile_map); i++) {
         if (qsv_profile_map[i].codec_id != codec_id)
             continue;
         if (qsv_profile_map[i].codec_profile == profile)
             return qsv_profile_map[i].mfx_profile;
+        if (avctx->hwaccel_flags & AV_HWACCEL_FLAG_ALLOW_PROFILE_MISMATCH &&
+            qsv_profile_map[i].codec_profile == FF_PROFILE_UNKNOWN ) {
+
+            av_log(avctx, AV_LOG_VERBOSE, "Codec %s profile %d not "
+                "supported for hardware decode.\n", codec_desc->name, profile);
+            av_log(avctx, AV_LOG_WARNING, "Using possibly-incompatible "
+                "mfx profile %d instead.\n", qsv_profile_map[i].mfx_profile);
+
+            return qsv_profile_map[i].mfx_profile;
+        }
```

Comment:
If this is really need to be fix, would better to fix in MSDK instead of ffmpeg-qsv. You will always set a main profile for unknown profile case but it may be main10 profile actually.

Reply:
> If this is really need to be fix, would better to fix in MSDK instead of ffmpeg-qsv. 
Since a specific flag AV_HWACCEL_FLAG_ALLOW_PROFILE_MISMATCH is set by user,  I thought it's acceptable to use a default value for hwaccel decode in this compatible mode. 
And there is a same path in vaapi, so I thought it will make sense.
https://github.com/FFmpeg/FFmpeg/blob/6c67c8ca9ae603b91e65274f93099ccb07bb33db/libavcodec/vaapi_decode.c#L479


>You will always set a main profile for unknown profile case but it may be main10 profile actually.

Yes, i've thought about this. And desc->comp[0].depth could be used to determine the possible profile (main or main10), and other desc information could also be parsed to guess the potential profile for HEVC_REXT, and it may be too complex for current stage.

Actually, after confirming with QA teams, I eliminate the depth relevant code because there is no profile 0 cases in HEVC 10 bit.

Reply:
It could be handled in MSDK in UMC_HEVC_DECODER::MFX_Utility::CheckVideoParam_H265():
https://github.com/Intel-Media-SDK/MediaSDK/blob/5b23796ab993771718fec3b997c3f4a5c98c6c4c/_studio/shared/umc/codec/h265_dec/src/umc_h265_mfx_utils.cpp#L987
But another question may be how can user control/know whether QSV uses a compatible profile in ffmpeg level?
Silently changing the profile in MSDK may not be good, and we need some feedback from MSDK.